src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/Extensions.cs

```diff
@@ -136,22 +136,76 @@ public static bool TryGetRegisterAttribute (this MethodDefinition method, out Cu
 		}
 
 		public static bool TryGetRegisterMember (this MethodDefinition md, out string method)
+		{
+			return TryGetRegisterMember (md, out method, out _, out _);
+		}
+
+		public static bool TryGetRegisterMember (this MethodDefinition md, out string method, out string nativeMethod, out string signature)
 		{
 			CustomAttribute register;
 			method = null;
+			nativeMethod = null;
+			signature = null;
 
 			if (!md.TryGetRegisterAttribute (out register))
 				return false;
 
 			if (register.ConstructorArguments.Count != 3)
 				return false;
 
+			nativeMethod = (string)register.ConstructorArguments [0].Value;
+			signature = (string)register.ConstructorArguments [1].Value;
 			method = (string)register.ConstructorArguments [2].Value;
 
 			if (string.IsNullOrEmpty (method))
 				return false;
 
 			return true;
 		}
+
+		static string MarshalMethodName (string name, string signature)
+		{
+                        var idx1 = signature.IndexOf ('(');
+                        var idx2 = signature.IndexOf (')');
+                        string arguments = null;
+
+                        if (idx1 >= 0 && idx2 >= idx1)
+                                arguments = signature.Substring (idx1 + 1, idx2 - idx1 - 1);
+
+                        if (string.IsNullOrEmpty (arguments))
+                                return null;
+
+                        return $"n_{name}_{arguments?.Replace ('/', '_')?.Replace (';', '_')}";
```

Comment:
I don't know what to make of this. From "context" -- the below `TryGetMarshalMethod()` body -- this will be used on a `__<$>_jni_marshal_methods` type to construct a method *on* the `__<$>_jni_marshal_methods` type.

I don't see how this will work, because this appears to suggest that the `__<$>_jni_marshal_methods` methods will be named following the pattern `$"n_{name}_{arguments}"`.

Compare with what Java.Interop produces:

```
$ make prepare all JI_MAX_JDK=8
$ make run-test-jnimarshal
$ cd bin/TestDebug/ && (MONO_PATH=. monop -r:Java.Interop.Export-Tests.dll 'Java.InteropTests.ExportTest+__<$>_jni_marshal_methods')
internal sealed class __<$>_jni_marshal_methods {
	
	public __<$>_jni_marshal_methods ();
	
	
	[Java.Interop.JniAddNativeMethodRegistrationAttribute()]
	public static void __RegisterNativeMembers (Java.Interop.JniNativeMethodRegistrationArguments args);
	public static IntPtr FuncIJavaObject (IntPtr __jnienv, IntPtr __this);
	public static long FuncInt64 (IntPtr __jnienv, IntPtr __this);
	public static void InstanceAction (IntPtr __jnienv, IntPtr __this);
	public static void StaticAction (IntPtr __jnienv, IntPtr __class);
	public static void StaticActionInt32String (IntPtr __jnienv, IntPtr __class, int i, IntPtr v);
	public static int StaticFuncMyLegacyColorMyColor_MyColor (IntPtr __jnienv, IntPtr __class, int color1, int color2);
	public virtual bool Equals (object obj);
	
	[System.Runtime.ConstrainedExecution.ReliabilityContractAttribute((System.Runtime.ConstrainedExecution.Consistency)3, (System.Runtime.ConstrainedExecution.Cer)2)]
	protected virtual void Finalize ();
	public virtual int GetHashCode ();
	public Type GetType ();
	protected object MemberwiseClone ();
	public virtual string ToString ();
}
```

The marshal methods *don't* have an `n_` prefix, nor do they have a suffix which encodes parameter types.

The `$"n_{name}_{arguments}"` pattern *is* (kinda, sorta) followed by *`generator`* output...in that there are methods with an `n_` prefix. However, the "arguments" "suffix" doesn't always exist; consider `src/Mono.Android/obj/Debug/android-27/mcw/Android.App.Activity.cs`:

```csharp
namespace Android.App {
	public partial class Activity {
		static IntPtr n_GetActionBar (IntPtr jnienv, IntPtr native__this);
		static void n_Dump_Ljava_lang_String_Ljava_io_FileDescriptor_Ljava_io_PrintWriter_arrayLjava_lang_String_ (IntPtr jnienv, IntPtr native__this, IntPtr native_prefix, IntPtr native_fd, IntPtr native_writer, IntPtr native_args);
	}
}
```

The `_ARGUMENTS` suffix only happens *when there are arguments*, which isn't the case for `n_GetActionBar()`. `$"n_{name}_{arguments}" allows for empty arguments, but *always* includes the second `_`, which it shouldn't.

Furthermore, there's a potential future problem here: our "encoding" scheme can result in *very* long method names, and [the C# compiler doesn't like it](https://bugzilla.xamarin.com/show_bug.cgi?id=48390). *A* way to fix that is to instead generate a *hash* of the argument list, e.g. instead of `n_Dump_Ljava_lang_String_Ljava_io_FileDescriptor_Ljava_io_PrintWriter_arrayLjava_lang_String_()` produce `n_Dump_b3c52818142fd6a15491e56d3d48b132()`. *If* `generator` changes in this manner, then `MarshalMethodName()` would likewise break -- *assuming* that it's supposed to be looking up "legacy" marshal method names, which I'm not entirely sure is a valid assumption.

Reply:
> I don't see how this will work, because this appears to suggest that the __<$>_jni_marshal_methods methods will be named following the pattern $"n_{name}_{arguments}".
>
> Compare with what Java.Interop produces:

In Java.Interop tests we use `JavaCallable` attribute and I kept the names same as before. For `Register` attribute I use `n_{name}_{arguments}`. This is indeed inspired by the generator and https://github.com/xamarin/xamarin-android/projects description of 'Improve Process Startup Time' project.

I am not sure about using the hashes, how would we handle a collision?

Reply:
I was thinking about that. `Java.Interop.MarshalMemberBuilder` sounds as a good place. Let also update the name then so that it doesn't contain the last `_` for methods without parameter(s) as you suggested.

Reply:
I was also wondering whether we should separate the name and arguments parts by a different character than `_`? I think method named `Method_I (int)` will collide with another method named `Method(int, int)`?

Reply:
Actually above mentioned methods would not collide, because the second one will have `n_method_II`. `Method_II ()` and `Method (int, int)` would I think.